File: FlutterPlugin/ios/Classes/DaroInterstitialAdManager.swift
import Flutter
import UIKit
import Daro

class DaroInterstitialAdManager: NSObject, FlutterStreamHandler {
    private var eventSink: FlutterEventSink?
    private var adLoaders: [String: DaroInterstitialAdLoader] = [:]
    private var ads: [String: DaroInterstitialAd] = [:]

    func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "loadAd":
            loadAd(call, result: result)
        case "isAdReady":
            isAdReady(call, result: result)
        case "showAd":
            showAd(call, result: result)
        case "destroyAd":
            destroyAd(call, result: result)
        default:
            result(FlutterMethodNotImplemented)
        }
    }

    private func extractAdUnitId(from call: FlutterMethodCall, result: @escaping FlutterResult) -> String? {
        guard let args = call.arguments as? [String: Any],
              let adUnitId = args["adUnitId"] as? String else {
            result(FlutterError(code: "INVALID_ARGS", message: "Missing adUnitId", details: nil))
            return nil
        }
        return adUnitId
    }

    private func loadAd(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let adUnitId = extractAdUnitId(from: call, result: result) else { return }

        let args = call.arguments as? [String: Any]
        let placement = args?["placement"] as? String

        let adUnit = DaroAdUnit(unitId: adUnitId)
        let loader = DaroInterstitialAdLoader(unit: adUnit)
        adLoaders[adUnitId] = loader

        loader.listener.onAdLoadSuccess = { [weak self] ad, adInfo in
            self?.ads[adUnitId] = ad

            ad.interstitialListener.onShown = { [weak self] adInfo in
                self?.sendAdInfoEvent(eventType: "onAdShown", adUnitId: adUnitId)
            }

            ad.interstitialListener.onFailedToShow = { [weak self] adInfo, error in
                self?.sendAdDisplayErrorEvent(eventType: "onAdFailedToShow", adUnitId: adUnitId, error: error)
            }

            ad.interstitialListener.onDismiss = { [weak self] adInfo in
                self?.sendAdInfoEvent(eventType: "onAdDismiss", adUnitId: adUnitId)
            }

            self?.sendAdInfoEvent(eventType: "onAdLoadSuccess", adUnitId: adUnitId)
        }

        loader.listener.onAdLoadFail = { [weak self] error in
            self?.sendAdLoadErrorEvent(eventType: "onAdLoadFail", adUnitId: adUnitId, error: error)
        }

        loader.listener.onAdClicked = { [weak self] adInfo in
            self?.sendAdInfoEvent(eventType: "onAdClicked", adUnitId: adUnitId)
        }

        loader.listener.onAdImpression = { [weak self] adInfo in
            self?.sendAdInfoEvent(eventType: "onAdImpression", adUnitId: adUnitId)
        }

        loader.loadAd()
        result(nil)
    }

    private func isAdReady(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let adUnitId = extractAdUnitId(from: call, result: result) else { return }

        let ready = ads[adUnitId] != nil
        result(ready)
    }

    private func showAd(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let adUnitId = extractAdUnitId(from: call, result: result) else { return }

        guard let ad = ads[adUnitId] else {
            result(FlutterError(code: "AD_NOT_LOADED", message: "Ad not loaded", details: nil))
            return
        }

        ad.show(viewController: nil)
        result(nil)
    }

    private func destroyAd(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let adUnitId = extractAdUnitId(from: call, result: result) else { return }

        adLoaders.removeValue(forKey: adUnitId)
        ads.removeValue(forKey: adUnitId)
        result(nil)
    }

    private func adInfoToMap(adUnitId: String) -> [String: Any] {
        return [
            "adUnitId": adUnitId,
            "format": "interstitial"
        ]
    }

    private func sendAdInfoEvent(eventType: String, adUnitId: String) {
        sendEvent([
            "eventType": eventType,
            "adUnitId": adUnitId,
            "adInfo": adInfoToMap(adUnitId: adUnitId)
        ])
    }

    private func sendAdLoadErrorEvent(eventType: String, adUnitId: String, error: DaroError) {
        sendEvent([
            "eventType": eventType,
            "adUnitId": adUnitId,
            "error": [
                "code": error.code.rawValue,
                "message": error.message,
                "adUnitId": adUnitId
            ]
        ])
    }

    private func sendAdDisplayErrorEvent(eventType: String, adUnitId: String, error: DaroError) {
        sendEvent([
            "eventType": eventType,
            "adUnitId": adUnitId,
            "error": [
                "code": error.code.rawValue,
                "message": error.message
            ]
        ])
    }

    private func sendEvent(_ event: [String: Any]) {
        DispatchQueue.main.async { [weak self] in
            self?.eventSink?(event)
        }
    }

    func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
        self.eventSink = events
        return nil
    }

    func onCancel(withArguments arguments: Any?) -> FlutterError? {
        self.eventSink = nil
        return nil
    }
}

File: FlutterPlugin/android/src/main/kotlin/so/daro/flutter/DaroInterstitialAdManager.kt
package so.daro.flutter

import android.app.Activity
import android.os.Handler
import android.os.Looper
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import droom.daro.core.adunit.DaroInterstitialAdUnit
import droom.daro.loader.DaroInterstitialAdLoader
import droom.daro.core.listener.DaroInterstitialAdLoaderListener
import droom.daro.core.model.DaroInterstitialAd
import droom.daro.core.listener.DaroInterstitialAdListener
import droom.daro.core.model.DaroAdInfo
import droom.daro.core.model.DaroAdLoadError
import droom.daro.core.model.DaroAdDisplayFailError

class DaroInterstitialAdManager : MethodChannel.MethodCallHandler, EventChannel.StreamHandler {
    private var activity: Activity? = null
    private var eventSink: EventChannel.EventSink? = null
    private val adLoaders = mutableMapOf<String, DaroInterstitialAdLoader>()
    private val ads = mutableMapOf<String, DaroInterstitialAd>()
    private val mainHandler = Handler(Looper.getMainLooper())

    fun setActivity(activity: Activity?) {
        this.activity = activity
    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "loadAd" -> loadAd(call, result)
            "isAdReady" -> isAdReady(call, result)
            "showAd" -> showAd(call, result)
            "destroyAd" -> destroyAd(call, result)
            else -> result.notImplemented()
        }
    }

    private fun extractAdUnitId(call: MethodCall, result: MethodChannel.Result): String? {
        val adUnitId = call.argument<String>("adUnitId")
        if (adUnitId == null) {
            result.error("INVALID_ARGS", "Missing adUnitId", null)
            return null
        }
        return adUnitId
    }

    private fun loadAd(call: MethodCall, result: MethodChannel.Result) {
        val adUnitId = extractAdUnitId(call, result) ?: return

        val placement = call.argument<String>("placement") ?: ""
        val activity = this.activity

        if (activity == null) {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }

        ads[adUnitId]?.destroy()
        ads.remove(adUnitId)
        adLoaders.remove(adUnitId)

        val adUnit = DaroInterstitialAdUnit(
            key = adUnitId,
            placement = placement
        )

        val loader = DaroInterstitialAdLoader(
            context = activity,
            adUnit = adUnit
        )
        adLoaders[adUnitId] = loader

        loader.setListener(createAdLoaderListener(adUnitId))
        loader.load()
        result.success(null)
    }

    private fun createAdLoaderListener(adUnitId: String): DaroInterstitialAdLoaderListener {
        return object : DaroInterstitialAdLoaderListener {
            override fun onAdLoadSuccess(ad: DaroInterstitialAd, adInfo: DaroAdInfo) {
                ads[adUnitId] = ad
                ad.setListener(createAdListener(adUnitId))
                sendAdInfoEvent("onAdLoadSuccess", adUnitId, adInfo)
            }

            override fun onAdLoadFail(err: DaroAdLoadError) {
                sendAdLoadErrorEvent("onAdLoadFail", adUnitId, err)
            }
        }
    }

    private fun createAdListener(adUnitId: String): DaroInterstitialAdListener {
        return object : DaroInterstitialAdListener {
            override fun onShown(adInfo: DaroAdInfo) {
                sendAdInfoEvent("onAdShown", adUnitId, adInfo)
            }

            override fun onFailedToShow(adInfo: DaroAdInfo, error: DaroAdDisplayFailError) {
                sendAdDisplayErrorEvent("onAdFailedToShow", adUnitId, error)
            }

            override fun onDismiss(adInfo: DaroAdInfo) {
                ads[adUnitId]?.destroy()
                ads.remove(adUnitId)
                sendAdInfoEvent("onAdDismiss", adUnitId, adInfo)
            }

            override fun onAdClicked(adInfo: DaroAdInfo) {
                sendAdInfoEvent("onAdClicked", adUnitId, adInfo)
            }

            override fun onAdImpression(adInfo: DaroAdInfo) {
                sendAdInfoEvent("onAdImpression", adUnitId, adInfo)
            }
        }
    }

    private fun isAdReady(call: MethodCall, result: MethodChannel.Result) {
        val adUnitId = extractAdUnitId(call, result) ?: return

        val ready = ads.containsKey(adUnitId)
        result.success(ready)
    }

    private fun showAd(call: MethodCall, result: MethodChannel.Result) {
        val adUnitId = extractAdUnitId(call, result) ?: return

        val ad = ads[adUnitId]
        if (ad == null) {
            result.error("AD_NOT_LOADED", "Ad not loaded", null)
            return
        }

        val activity = this.activity
        if (activity == null) {
            result.error("NO_ACTIVITY", "Activity not available", null)
            return
        }

        ad.show(activity = activity)
        result.success(null)
    }

    private fun destroyAd(call: MethodCall, result: MethodChannel.Result) {
        val adUnitId = extractAdUnitId(call, result) ?: return

        ads[adUnitId]?.destroy()
        ads.remove(adUnitId)
        adLoaders.remove(adUnitId)
        result.success(null)
    }

    private fun adInfoToMap(adInfo: DaroAdInfo, adUnitId: String): Map<String, Any> {
        return mapOf(
            "adUnitId" to adUnitId
        )
    }

    private fun sendAdInfoEvent(eventType: String, adUnitId: String, adInfo: DaroAdInfo) {
        sendEvent(
            mapOf(
                "eventType" to eventType,
                "adUnitId" to adUnitId,
                "adInfo" to adInfoToMap(adInfo, adUnitId)
            )
        )
    }

    private fun sendAdLoadErrorEvent(eventType: String, adUnitId: String, error: DaroAdLoadError) {
        sendEvent(
            mapOf(
                "eventType" to eventType,
                "adUnitId" to adUnitId,
                "error" to mapOf(
                    "code" to error.code,
                    "message" to error.message,
                    "adUnitId" to adUnitId
                )
            )
        )
    }

    private fun sendAdDisplayErrorEvent(eventType: String, adUnitId: String, error: DaroAdDisplayFailError) {
        sendEvent(
            mapOf(
                "eventType" to eventType,
                "adUnitId" to adUnitId,
                "error" to mapOf(
                    "code" to -1,
                    "message" to error.message
                )
            )
        )
    }

    private fun sendEvent(event: Map<String, Any>) {
        mainHandler.post {
            eventSink?.success(event)
        }
    }

    override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
        this.eventSink = events
    }

    override fun onCancel(arguments: Any?) {
        this.eventSink = null
    }
}
